### DISCUSSION TITLE
Send Value on MouseUp -- solved

### DISCUSSION THREAD

**Discussion by Godlike-Productions** [RomanKubiak/ctrlr#228] [Q&A] [answered]
Does anyone know how to determine if the mouse is down on a slider?

On the panel I'm working on there is an AlphaWheel.

The Kurzweil has a function to send Sysex for the number of clicks to send (in a single message), so that you can send a single sysex message for say 15 clicks, rather than sending 15 sysex messages.

We have a uiImageSlider that springs to 64.  We want to send value - 64 in this sysex message when we release the mouseButton

Anyone know how to do this?  

The methods here - https://ctrlr.org/forums/topic/detect-mouse-down-lua/ don't work, and I can't get isMouseButtonDown, mouseDown, mouseUp to return anything.

similarly getProperty("isMouseButtonDown") doesn't return anything.

Otherwise, does anyone know how to use a slider resources in the CustomComponent?

**Comment by Godlike-Productions** [comment, marked as answer]
I think I found it.

You can check with event.mods:isLeftButtonDown() if you are using on of the mouse driven events (like drag).  Now to make it do something when I stop dragging.  I wish onMouseUp was implemented for components other than CustomComponents

**Reply by Godlike-Productions** [reply]
Unfortunately the MouseDrag exits before the event can determine if the button is up, but this opened up another possibility.

I found a sneaky way around it thought. This will ONLY work on controllers that spring back to a value (in this case 0)

set a method to componentLuaMouseDrag

Code is
```
alphaWheelMouseDragMethod = function(--[[ CtrlrComponent --]] comp, --[[ MouseEvent --]] event)

-- mouseState = event.mods:isLeftButtonDown() -- this shows how to monitor LeftButton is down
-- amount = event:getDistanceFromDragStartY() -- this shows how to calculate Y distance
--panel:setGlobalVariable(0,amount) -- this shows how to set a global variable.  You can call a method if this changes

-- declared in panelLoaded -- alphaWheel = panel:getModulatorByName("alphaWheel")

tempValue = alphaWheel:getValue() -- this is the only line you need

end
```

next assign a method to the same modulator under luaModulatorValueChange.

The code for this is
```
alphaWheelValueChangedMethod = function(--[[ CtrlrModulator --]] mod, --[[ number --]] value, --[[ number --]] source)

if value == 0 then
    if tempValue ~= 0 then
        console (String("value: "..tempValue))
        -- put the code you want in here. In my case I'll send a sysex string
    end --if
end --if

end
```

When it springs back to 0 tempValue won't be updated to 0 because it is in the drag method and will remain at the final value that you dragged the wheel to.  tempValue must be global.

**Reply by dnaldoog** [reply]
How do you get a controller to spring back to 0 like uiSlider etc?

**Reply by Godlike-Productions** [reply]
It is a uiSlider (well a uiImageSlider).  We have spring turned on, and the spring value of 0.

The alphaWheel is the big black knob to the right of the LCD display.

![2021-03-28](https://user-images.githubusercontent.com/13851825/112746883-857fe900-8fe4-11eb-9d2d-ed606cd2bebf.png)

**Reply by dnaldoog** [reply]
Crazy! I never noticed that feature before. Thank you. Nice looking panel!

**Reply by Godlike-Productions** [reply]
The good looks are all courtesy of bijlevel. I'm trying to make the backend look as pretty as the front panel. :P

**Reply by Ted-juh** [reply]
Nice workaround!!

Still a question:
> --panel:setGlobalVariable(0,amount) -- this shows how to set a global variable.  You can call a method if this changes

I'm confused. Is this a Ctrlr-only thing? I'm asking because I can't find anything about it in the Juce classes. And what does it do? Does it put all the "global" variable names in a table/ memory like when you declare modulators in a init method after Panelloaded? Is there any advantage of doing it this way instead of not declaring it "local"?

And what do you mean by "You can call a method if this changes"? 

Last question: 
Is that where "panelGlobalVariables" with all the -1: -1: -1: -1 in the general properties tab is for?

Edit: By the time I wrote this question, several replies were done. I really need to learn to type faster. Reason for the edit: Nice looking panel indeed. That's some skeuomorphism design I can get really jealous off. Big thumbs up.

**Reply by Godlike-Productions** [reply]
Yeah the GlobalVariables are a ctrlr thing I think, though I've seen them referred to with other LUA uses.  It may be a place to pass globals between LUA and the host. Not really sure.  I think they were included mainly for use for people creating panels using the toolbar, without LUA, but there is a callback for them.  When one changes you can call a method, which is a handy alternative to timers in many situations.  Yeah that list is the globalVariables.  they start at k0 and each one after the ; is a new one.  You can use the global variables in the SYSEX field from the front panel by puting k0, k1, k2 in, etc.

**Reply by Godlike-Productions** [reply]
My favourite part of this panel is that the LCD screen is dynamic.  It updates to reflect what's on the front panel of the synth.  I have a bit more functionality I want to build into this. I'll see how far down the rabbit hole I go.

**Reply by dnaldoog** [reply]
Here's a panel that can explain it!
[global warming_1_1_a_2021-03-28_18-26.zip](https://github.com/RomanKubiak/ctrlr/files/6217076/global.warming_1_1_a_2021-03-28_18-26.zip)

**Reply by dnaldoog** [reply]
> Is that where "panelGlobalVariables" with all the -1: -1: -1: -1 in the general properties tab is for?

Yes each one represents a potential global variable.

**Reply by dnaldoog** [reply]
> My favourite part of this panel is that the LCD screen is dynamic. It updates to reflect what's on the front panel of the synth. I have a bit more functionality I want to build into this. I'll see how far down the rabbit hole I go.

Did you ever see this post?

https://ctrlr.org/forums/topic/creating-images-using-lua/#post-120273

**Reply by Godlike-Productions** [reply]
Yep, we certainly did. It did help us.

Basically bijlevel filled a rectangle and then changes the colour of each pixel as it's read from the sysex dump if it is active.  

It's a balancing act getting the text to fit to the graphic, as they are seperate objects.  The graphics layer is off in that image, as I'm working offline, not connected to the Kurz.  MIDI is a bit slow over a VPN from 350km away.

**Reply by dnaldoog** [reply]
I can imagine, especially if kangaroos are stomping on the cables! HaHa - Will be very interested to see how you implement that XPM type code if you do use it that is!

**Reply by Godlike-Productions** [reply]
We just drew little 1 pixel rectangles over the top of the big rectangle.  It's really a tiny graphic that has been stretched.

```
if p == 1 then
            	g:setColour(Colour(0xffffffff)) -- white
		g:fillRect(x, y, 1, 1)           
end
```

**Reply by dobo365** [reply]
Excellent tip about the spring! I also didn't know that.
Adding to my step by step guide ;-)
